plugins/out_stackdriver/stackdriver.c

```diff
@@ -277,6 +277,170 @@ static char *get_google_token(struct flb_stackdriver *ctx)
     return ctx->o->access_token;
 }
 
+static bool validate_msgpack_unpacked_data(msgpack_object root)
+{
+    return root.type == MSGPACK_OBJECT_ARRAY &&
+           root.via.array.size == 2 &&
+           root.via.array.ptr[1].type == MSGPACK_OBJECT_MAP;
+}
+
+static char *get_str_value_from_msgpack_map(msgpack_object_map map, 
+                                            const char *key)
+{
+    int i;
+    int key_size;
+    msgpack_object k;
+    msgpack_object v;
+    char *ptr = NULL;
+
+    key_size = strlen(key);
+    for (i = 0; i < map.size; i++) {
+        k = map.ptr[i].key;
+        v = map.ptr[i].val;
+
+        if (k.type != MSGPACK_OBJECT_STR) {
+            continue;
+        }
+
+        if (k.via.str.size == key_size && 
+            strncmp(key, (char *) k.via.str.ptr, k.via.str.size) == 0) {
+            /* make sure to free it after use */ 
+            ptr =  flb_strndup(v.via.str.ptr, v.via.str.size);
+            break;
+        }
+    }
+
+    return ptr;
+}
+
+/* 
+*    process_local_resource_id():
+*  - extract the value from "logging.googleapis.com/local_resource_id" field in the record
+*  - use extracted value to assign the label keys for differnet resource types that specified
+*    in the configuration of stackdriver_out plugin
+*/
+static int process_local_resource_id(const void *data, size_t bytes,
+                                     struct flb_stackdriver *ctx, char *type)
+{
+    int len_ptr;
+    int ret = -1;
+    size_t off = 0;
+    msgpack_object root;
+    msgpack_object_map map;
+    msgpack_unpacked result;
+    char *local_resource_id;
+    char *ptr;
+    const char delim[] = ".";
+
+    msgpack_unpacked_init(&result);
+    if (msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS) {
+        root = result.data;
+
+        if (!validate_msgpack_unpacked_data(root)) {
+            flb_plg_warn(ctx->ins, "unexpected record format");
+            return -1;
+        }
+
+        map = root.via.array.ptr[1].via.map;
+        local_resource_id = get_str_value_from_msgpack_map(map, LOCAL_RESOURCE_ID_KEY);
+        if (!local_resource_id) {
+            msgpack_unpacked_destroy(&result);
+            return -1;
+        }
+        
+        ptr = strtok(local_resource_id, delim);
+        len_ptr = strlen(ptr);
+        if (strncmp(type, "k8s_container", strlen("k8s_container")) == 0) {
+            /* check the prefix */
+            if (strncasecmp(ptr, type, len_ptr) != 0) {
+                free(local_resource_id);
+                msgpack_unpacked_destroy(&result);
+                return -1;
+            }
+
+            ptr = strtok(NULL, delim);
+            while (ptr != NULL)
```

Comment:
align the brace on the right side:

https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

Reply:
updated